keps/0012-label-index.md

```diff
@@ -0,0 +1,51 @@
+---
+kep-number: 12
+title: Index labels to support efficient queries by label selectors
+authors:
+  - "@jian-he, @ccding"
+owning-sig: sig-api-machinery
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: "2018-05-31"
+last-updated: "2018-05-31"
+status: provisional
+---
+
+## Summary
+
+This proposal presents solutions to make the queries by label selectors in the
+list API efficient.
+
+## Motivation
+
+Currently, only namespace index is supported in the cache. Controllers such as
+ReplicaSet, when selecting pods, need to select all the pods by namespace and
+then select the ones wanted filtering by index. While running a large number
+of pods, we observed that listing all pods and then filtering takes several
+seconds. This proposal presents solutions to improve this.
```

Comment:
Can you describe the scenario you are hitting?  With the watch cache enabled I have never seen a filtered label query take longer than 1s even on large sets, so I’d want to understand the data you have and how it could be applied.

Reply:
Eventually we will have to do this. However, I don't want to do it too early because it's a lot of complexity, and I'm not sure apiserver is the right place to do it. I think it makes more sense to have the storage layer do this. This requires either adding a layer between apiserver and etcd, or implementing a bunch of features in etcd. Neither is going to happen soon. In the meantime, I don't want to do this in apiserver unless it's absolutely necessary.

Reply:
Added evaluation results. PTAL

Reply:
@lavalamp Adding a layer between apiserver and etcd sounds like a reasonable approach, build DB like index query into etcd seems like not that possible. So this KEP just focus on client side.